SimG4CMS/Calo/src/HFFibre.cc

```diff
@@ -12,86 +12,108 @@
 #include <sstream>
 
 //#define EDM_ML_DEBUG
+HFFibre::Params::Params(double iFractionOfSpeedOfLightInFibre,
+                        const HcalDDDSimConstants* hcons,
+                        const HcalSimulationParameters* hps)
+    : fractionOfSpeedOfLightInFibre_{iFractionOfSpeedOfLightInFibre},
+      gParHF_{hcons->getGparHF()},
+      rTableHF_{hcons->getRTableHF()},
+      shortFibreLength_{hps->shortFiberLength_},
+      longFibreLength_{hps->longFiberLength_},
+      attenuationLength_{hps->attenuationLength_},
+      lambdaLimits_{{static_cast<double>(hps->lambdaLimits_[0]), static_cast<double>(hps->lambdaLimits_[1])}} {}
 
-HFFibre::HFFibre(const std::string& name,
-                 const HcalDDDSimConstants* hcons,
-                 const HcalSimulationParameters* hps,
-                 edm::ParameterSet const& p)
-    : hcalConstant_(hcons), hcalsimpar_(hps) {
-  edm::ParameterSet m_HF =
-      (p.getParameter<edm::ParameterSet>("HFShower")).getParameter<edm::ParameterSet>("HFShowerBlock");
-  cFibre = c_light * (m_HF.getParameter<double>("CFibre"));
-  edm::LogVerbatim("HFShower") << "HFFibre:: Speed of light in fibre " << cFibre / (CLHEP::m / CLHEP::ns) << " m/ns";
+HFFibre::HFFibre(const HcalDDDSimConstants* hcons, const HcalSimulationParameters* hps, edm::ParameterSet const& p)
+    : HFFibre(Params(p.getParameter<edm::ParameterSet>("HFShower")
+                         .getParameter<edm::ParameterSet>("HFShowerBlock")
+                         .getParameter<double>("CFibre"),
+                     hcons,
+                     hps)) {}
+
+HFFibre::HFFibre(Params iP)
+    : cFibre_(c_light * iP.fractionOfSpeedOfLightInFibre_),
+      gpar_(std::move(iP.gParHF_)),
+      radius_(std::move(iP.rTableHF_)),
+      shortFL_(std::move(iP.shortFibreLength_)),
+      longFL_(std::move(iP.longFibreLength_)),
+      attL_(std::move(iP.attenuationLength_)),
+      lambLim_(iP.lambdaLimits_) {
+  edm::LogVerbatim("HFShower") << "HFFibre:: Speed of light in fibre " << cFibre_ / (CLHEP::m / CLHEP::ns) << " m/ns";
   // Attenuation length
-  attL = hcalsimpar_->attenuationLength_;
-  nBinAtt = static_cast<int>(attL.size());
-  std::stringstream ss1;
-  for (int it = 0; it < nBinAtt; it++) {
-    if (it / 10 * 10 == it) {
-      ss1 << "\n";
+  nBinAtt_ = static_cast<int>(attL_.size());
+
+  edm::LogVerbatim("HFShower").log([&](auto& logger) {
```

Comment:
`LogVerbatim::log` will only call the lambda if the message will potentially be sent. This avoids building the `std::stringstream` only to have its contents thrown away.